ably/realtime_conn_spec_test.go

```diff
@@ -183,6 +183,245 @@ func connectAndWait(t *testing.T, realtime *ably.Realtime) {
+		}
+
+		// Expect timer for a message receive.
+		var timer ablytest.AfterCall
+
+		ablytest.Instantly.Recv(t, &timer, afterCalls, t.Fatalf)
+		if expected, got := receiveTimeout, timer.D; expected != got {
+			t.Fatalf("expected %v, got %v", expected, got)
+		}
+
+		// Let the deadline pass without a message; expect a disconnection.
+		timer.Fire()
+
+		ablytest.Soon.Recv(t, &change, stateChange, t.Fatalf)
+		if expected, got := ably.ConnectionStateClosed, change.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change.Current)
+		}
+
+		errReason := change.Reason
+		// make sure the reason is timeout
+		if !strings.Contains(errReason.Error(), "timeout") {
+			t.Errorf("expected %q to contain timeout", errReason.Error())
+		}
+
+		ablytest.Instantly.NoRecv(t, nil, stateChange, t.Fatalf)
+	})
+
+	t.Run("RTN12c: transition to closed on transport error", func(t *testing.T) {
+		app, client, doEOF := setUpWithEOF()
+		defer safeclose(t, ablytest.FullRealtimeCloser(client), app)
+
+		stateChange := make(connectionStateChanges, 2)
+		client.Connection.OnAll(stateChange.Receive)
+
+		client.Close()
+
+		var change ably.ConnectionStateChange
+		ablytest.Soon.Recv(t, &change, stateChange, t.Fatalf)
+
+		if expected, got := ably.ConnectionStateClosing, change.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change.Current)
+		}
+
+		doEOF <- struct{}{}
+
+		select {
+		case change = <-stateChange:
+		case <-time.After(50 * time.Millisecond):
+			t.Fatal("didn't transition on EOF")
+		}
+
+		if expected, got := ably.ConnectionStateClosed, change; expected != got.Current {
+			t.Fatalf("expected transition to %v, got %v", expected, got)
+		}
+
+		ablytest.Instantly.NoRecv(t, nil, stateChange, t.Fatalf)
+	})
+
+	connectionStates := [2]ably.ConnectionState{ably.ConnectionStateDisconnected, ably.ConnectionStateSuspended}
+	for _, connectionState := range connectionStates {
+		t.Run(fmt.Sprintf("RTN12d : Should directly close on %v", connectionState), func(t *testing.T) {
+			app, client, _ := setUpWithEOF()
+			defer safeclose(t, ablytest.FullRealtimeCloser(client), app)
+
+			client.Connection.SetState(connectionState, nil, time.Minute)
+			stateChange := make(connectionStateChanges, 1)
+			client.Connection.OnAll(stateChange.Receive)
+			client.Connection.Close()
+
+			var change ably.ConnectionStateChange
+			ablytest.Soon.Recv(t, &change, stateChange, t.Fatalf)
+
+			if expected, got := ably.ConnectionStateClosed, change.Current; expected != got {
+				t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change.Current)
+			}
+			ablytest.Instantly.NoRecv(t, nil, stateChange, t.Fatalf)
+		})
+	}
+
+	t.Skip("RTN12f: transition to closed when close is called intermittently")
+	// todo - Need to fix the test and check for possible race conditions
+	t.Run("RTN12f: transition to closed when close is called intermittently", func(t *testing.T) {
```

Comment:
@tcard you can check the test, I don't think I am missing anything here. Not sure why still it's not working : (